### PR TITLE
Adding flexible input sources for Diarization Mixin

### PR DESCRIPTION
**Pull request by tango4j**
# What does this PR do ?

Reflecting the change requested by:

[Issues #14495](https://github.com/NVIDIA-NeMo/NeMo/issues/14495) 
[Issues #14977](https://github.com/NVIDIA-NeMo/NeMo/issues/14977) 

- Force end-to-end diarization to use the first channel audio if it is multichannel audio.
- Adding numpy input format to end-to-end speaker diarization (Sortformer models) 
- Added offset/duration masking feature to `.diarize()` (SpkDiarizationMixin) function
- Added max-speaker limitation to multi-speaker inference pipeline
- Revamped the diarization mixin unit-tests.

**Collection**: [Note which collection this PR will affect]

# Changelog

The following files are changed.

```bash
nemo/collections/asr/data/audio_to_diar_label_lhotse.py                                                                                                                                                                                                                             
nemo/collections/asr/models/sortformer_diar_models.py                                                                                                                                                                                                                               
nemo/collections/asr/parts/mixins/diarization.py                                                                                                                                                                                                                                    
nemo/collections/asr/parts/utils/multispk_transcribe_utils.py                                                                                                                                                                                                                       
tests/collections/speaker_tasks/mixins/test_diarization.py
```

# Usage

- You can potentially add a usage example below

```python
from nemo.collections.asr.models import SortformerEncLabelModel

# load model from Hugging Face model card directly (You need a Hugging Face token)
diar_model = SortformerEncLabelModel.from_pretrained("nvidia/diar_streaming_sortformer_4spk-v2")

# If you have a downloaded model in "/path/to/diar_streaming_sortformer_4spk-v2.nemo", load model from a downloaded file
diar_model = SortformerEncLabelModel.restore_from(restore_path="/path/to/diar_streaming_sortformer_4spk-v2.nemo", map_location='cuda', strict=False)

# switch to inference mode
diar_model.eval()

predicted_segments = diar_model.diarize(audio=audio_input, batch_size=1)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in NeMo ASR

# Additional Information

Related issues: 
[Issues #14495](https://github.com/NVIDIA-NeMo/NeMo/issues/14495) - A request for numpy input for `.diarize()` function.
[Issues #14977](https://github.com/NVIDIA-NeMo/NeMo/issues/14977) - Duration is not picked up when manifest is used for `.diarize()` function.

### CODE REVIEW

File: nemo/collections/asr/data/audio_to_diar_label_lhotse.py [hunk 1]
```diff
@@ -58,6 +58,8 @@ def __init__(self, cfg):
         self.num_mel_frame_per_target_frame = int(self.cfg.get('subsampling_factor', 8))
 
     def __getitem__(self, cuts) -> Tuple[torch.Tensor, ...]:
+        # NOTE: This end-to-end diarization dataloader only loads the 1st ch of the audio file.
```
Comment (KunalDhawan) [inline nemo/collections/asr/data/audio_to_diar_label_lhotse.py:61]: Should we add log this somewhere so that users don't run into unexpected issues?

**Review by KunalDhawan — APPROVED** [review]
LGTM, great work @tango4j!

File: nemo/collections/asr/data/audio_to_diar_label_lhotse.py (hunk 1 above)
Comment (tango4j) [reply nemo/collections/asr/data/audio_to_diar_label_lhotse.py:61]: Added the warning message and made these into one for loop.

File: nemo/collections/asr/models/sortformer_diar_models.py
Comment (ooobo) [inline nemo/collections/asr/models/sortformer_diar_models.py:1145]: do we need sample_rate here too?

Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py:1145]: Yes because if numpy input is assigned to `audio`, we do not know what sample rate `audio` time series signal is based on. This is mentioned in the Huggingface README. default is 16000 always.

Comment (ooobo) [reply nemo/collections/asr/models/sortformer_diar_models.py:1145]: Sorry I wasn't quite clear, I'm wondering if we need to return sample_rate here too.
Running the example as per the readme fails when I try it, as sample_rate is not passed through:
```python
from nemo.collections.asr.models import SortformerEncLabelModel
diar_model = SortformerEncLabelModel.from_pretrained("nvidia/diar_streaming_sortformer_4spk-v2")
diar_model.eval()

import numpy as np
# Create dummy audio: 10 seconds of white noise at 16kHz
audio_input = np.random.randn(16000 * 10).astype(np.float32)

diar_model.diarize(audio=audio_input, batch_size=1, sample_rate=16000)

Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py:1145]: You are correct and I made a mistake. 
sample_rate variable is missing in `super().diarize()` function.
I checked and I missed this part. 
I will make a PR that fixes that.

Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py:1145]: The diarize() function in this file
`NeMo/nemo/collections/asr/parts/mixins/diarization.py`
has "sample_rate" so this fix will do the job.

Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py:1145]: @ooobo 
Thank you so much for pointing this out.
I prepared a PR for this: https://github.com/NVIDIA-NeMo/NeMo/pull/15228

Comment (ooobo) [reply nemo/collections/asr/models/sortformer_diar_models.py:1145]: Perfect, all set now
Appreciate all your hard work!
